apps/agentic-chat/src/lib/langchain.ts

```diff
@@ -0,0 +1,58 @@
+import { ChatOpenAI } from '@langchain/openai';
+import { HumanMessage, AIMessage } from '@langchain/core/messages';
+import { ToolNode } from '@langchain/langgraph/prebuilt';
+import {
+  StateGraph,
+  MessagesAnnotation,
+  START,
+  END,
+} from '@langchain/langgraph/web';
+import { tokensSearch, bebopRate } from '@agentic-chat/tools';
+
+// Define the tools for the agent to use
+const tools = [tokensSearch, bebopRate];
+const toolNode = new ToolNode(tools);
+
+// Create a model and give it access to the tools
+const model = new ChatOpenAI({
+  model: 'gpt-4o-mini',
+  temperature: 0,
+  openAIApiKey: import.meta.env.VITE_OPENAI_API_KEY,
+}).bindTools(tools);
+
+// Define the function that determines whether to continue or not
+function shouldContinue({ messages }: typeof MessagesAnnotation.State) {
+  const lastMessage = messages[messages.length - 1] as AIMessage;
+
+  // If the LLM makes a tool call, then we route to the "tools" node
+  if (lastMessage.tool_calls?.length) {
+    return 'tools';
+  }
+  // Otherwise, we stop (reply to the user) using the special "__end__" node
+  return END;
+}
+
+// Define the function that calls the model
+async function callModel(state: typeof MessagesAnnotation.State) {
+  const response = await model.invoke(state.messages);
+  return { messages: [response] };
+}
+
+// Define a new graph
+const workflow = new StateGraph(MessagesAnnotation)
+  .addNode('agent', callModel)
+  .addEdge(START, 'agent')
+  .addNode('tools', toolNode)
+  .addEdge('tools', 'agent')
+  .addConditionalEdges('agent', shouldContinue);
```

Comment:
For the visual learners.

<img width="569" alt="Screenshot 2025-05-10 at 12 13 42 pm" src="https://github.com/user-attachments/assets/c7ee076c-14ca-4e76-b37e-7c2dbc1d4aca" />

Reply:
One thing to note here @0xApotheosis - langchain project and the webapp do work a bit differently because of the use of the (deprecated) system prompts in langchain projecs mostly, I haven't really gone back to it on the follow-up PR which wires up things a bit more, but it still exists for our convenience.
Not sure if we'll end up using the langchain project in the end (i.e langsmith app + https://agentchat.vercel.app/) but if we do, we'll probably want to ensure that we don't go with langgraph ReAct flavour there (reason + act) but rather go with the legacy langchain agents flow i.e the same used in the webapp, to ensure that the two models work similarly.